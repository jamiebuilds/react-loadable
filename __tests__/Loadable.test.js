// @noflow
declare var test: any;
import path from "path";
import React from "react";
import renderer from "react-test-renderer";
import Loadable from "../src";

let waitFor = (delay: number) => {
  return new Promise(resolve => {
    setTimeout(resolve, delay);
  });
};

let createLoader = (delay, Component, error?) => {
  return async () => {
    await waitFor(delay);
    if (Component) {
      return Component;
    } else {
      throw error;
    }
  };
};

let MyLoadingComponent = props => (
  <div>MyLoadingComponent {JSON.stringify(props)}</div>
);
let MyComponent = props => <div>MyComponent {JSON.stringify(props)}</div>;

test("loading success", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, MyComponent),
    MyLoadingComponent
  );

  let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded

  let component2 = renderer.create(<LoadableMyComponent prop="bar" />);

  expect(component2.toJSON()).toMatchSnapshot(); // reload
});

test("loading error", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, null, new Error("test error")),
    MyLoadingComponent
  );

  let component = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // errored
});

test("server side rendering", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, null, new Error("test error")),
    MyLoadingComponent,
    null,
    path.join(__dirname, "../__fixtures__/component.js")
  );

  let component = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component.toJSON()).toMatchSnapshot(); // serverside
});

test("server side rendering es6", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, null, new Error("test error")),
    MyLoadingComponent,
    null,
    path.join(__dirname, "../__fixtures__/component.es6.js")
  );

  let component = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component.toJSON()).toMatchSnapshot(); // serverside
});

test("preload", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, MyComponent),
    MyLoadingComponent,
    null
  );

  LoadableMyComponent.preload();
  await waitFor(200);

  let component1 = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component1.toJSON()).toMatchSnapshot(); // still loading...
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // success

  let component2 = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component2.toJSON()).toMatchSnapshot(); // success
});
