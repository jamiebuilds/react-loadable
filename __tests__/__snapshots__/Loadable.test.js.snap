// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading error 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"error":null}
</div>
`;

exports[`loading error 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"error":null}
</div>
`;

exports[`loading error 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":false,"pastDelay":false,"error":{}}
</div>
`;

exports[`loading success 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"error":null}
</div>
`;

exports[`loading success 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"error":null}
</div>
`;

exports[`loading success 3`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`loading success 4`] = `
<div>
  MyComponent 
  {"prop":"bar"}
</div>
`;

exports[`preload 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"error":null}
</div>
`;

exports[`preload 2`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`preload 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`resolveModule 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"error":null}
</div>
`;

exports[`resolveModule 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"error":null}
</div>
`;

exports[`resolveModule 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`server side rendering 1`] = `
<div>
  fixture1
</div>
`;

exports[`server side rendering es6 1`] = `
<div>
  fixture2
</div>
`;
